src/vdaf.rs

```diff
@@ -149,18 +143,17 @@ where
     type OutputShare: Clone + Debug;
 
     /// An Aggregator's share of the aggregate result.
-    type AggregateShare: Aggregatable<OutputShare = Self::OutputShare> + for<'a> TryFrom<&'a [u8]>;
+    type AggregateShare: Aggregatable<OutputShare = Self::OutputShare>
+        + TryFrom<Vec<u8>>
```

Comment:
I believe your claim that the copy would be optimized away, but that aside, I think the ergonomics of `TryFrom<Vec<u8>>` are worse than `for<'a> TryFrom<&'a [u8]>`, regardless of performance. I also get the argument that `TryFrom` should consume the input value when doing the conversion, but in the stdlib alone there are several `From` implementations that seem analogous to this one, like [`From<&str> for Vec<u8>`](https://doc.rust-lang.org/std/convert/trait.From.html#impl-From%3C%26str%3E-for-Vec%3Cu8%2C%20Global%3E).

I agree with you that there should be no or little impact from the copy forced by taking a `Vec<u8>` by value, so I can live with either trait bound.

Reply:
Interested in folks' thoughts on the ergonomics -- IMO, dropping the infectious where clause is so nice that almost any other changes in ergonomics are acceptable. (Could you clarify what parts of the ergonomics you dislike here, possibly I'm missing something bad.)

That said, using `From`/`Into` is not the only way to phrase this change. For example, we could introduce a custom trait that provides equivalent functionality without the need to worry about how smart the compiler is. Something like:

```rust
trait Serializable {
  fn serialize(&self) -> Vec<u8>;
  fn deserialize(buf: &[u8]) -> Result<Self, SomeError>;
}
```

...except that we'd probably want to pick different names to avoid confusion with `serde`'s serialization traits. A custom trait also has the downside that stdlib & other libraries are not going to know how to work with it.

Reply:
I would prefer separate traits for this serialization/deserialization. We probably shouldn't reuse `Encode`/`Decode`, as here we rely on greedy "read to end" behavior, unlike everything in the codec module.

Reply:
(another approach would be to have `AggregateShare` implement `Into<Vec<u8>>` and `for<'a> TryFrom<&'a [u8]>`. this would drop the infectious where clause while still allowing parsing of arbitrary slices of bytes.)

Reply:
The bad ergonomics of `TryFrom<Vec<u8>>` are that callers who have a `&[u8]` are forced to do something like `to_vec`, but if we required `for<'a> TryFrom<&'a [u8]>`, then it's easy for callers who have a `Vec<u8>` to pass a reference to it. So I think we're on the same page: `AggregateShare` requiring `Into<Vec<u8>>` and `for<'a> TryFrom<&'a [u8]>` gets us all the wins we want.

Reply:
OK -- I rephrased things in terms of `for<'a> TryFrom<&'a [u8]>`. This change also let me drop the `F: FieldElement` requirement on the `From<Vec<F>>` implementations, too.

While I was at it, I added trait requirements on `OutputShare` equivalent to `AggregateShare` -- they were already implemented (and tested), and are logically required to run a VDAF evaluation (assuming the different parties are separated by an IPC boundary), so I think `OutputShare` is equally justified in requiring these traits as `AggregateShare` is.